### ISSUE TITLE
Crash when using JWT as password

### ISSUE THREAD

**Issue by sschimmel** [processone/ejabberd#3067]
## Environment

- ejabberd version: 19.09
- Erlang version: `erl +V`
- OS: Linux (Ubuntu)
- Installed from: official binary installer

## Errors from error.log/crash.log

```
07:56:17.251 [info] (<0.544.0>) Accepted connection [::ffff:138.201.91.94]:36194 -> [::ffff:10.18.0.8]:5222
07:56:17.552 [error] gen_server <0.544.0> terminated with reason: no function clause matching jose_jws_alg_rsa_pkcs1_v1_5:verify({jose_jwk,{jose_jwk_set,[{jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPrivateKey','two-prime',21742263746488030104405521343183166191161927668668832678963882533224849627713697096673808942396166632655001363829479887991333419338499736733930419606265074138668507628098162744273934555101659209242746136159013755169313875220443887183686817460947357037265862949520061234764593811698354537662381041160306313905660977482112950644760943819046932440697293474045016208686816396415168146427032289769084327339610569003142494746410319845820075666726582865656725406634768508632406115803852827347182888089585019913453855785437574949464529944685116786371925939171960320238722474399781172007509383004790996974864098331358310421957,...}},...}]},...}, <<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJqaWQiOiJTZWJhc3RpYWFuQGNoYXQuZGV2LnJlenpvbmF0aW9uLmFwcC...">>, <<12,80,80,93,77,192,82,118,51,227,93,211,133,237,4,49,42,143,3,232,59,74,4,147,89,105,118,93,232,...>>, 'RS256') line 64
07:56:17.553 [error] CRASH REPORT Process <0.544.0> with 0 neighbours exited with reason: no function clause matching jose_jws_alg_rsa_pkcs1_v1_5:verify({jose_jwk,{jose_jwk_set,[{jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPrivateKey','two-prime',21742263746488030104405521343183166191161927668668832678963882533224849627713697096673808942396166632655001363829479887991333419338499736733930419606265074138668507628098162744273934555101659209242746136159013755169313875220443887183686817460947357037265862949520061234764593811698354537662381041160306313905660977482112950644760943819046932440697293474045016208686816396415168146427032289769084327339610569003142494746410319845820075666726582865656725406634768508632406115803852827347182888089585019913453855785437574949464529944685116786371925939171960320238722474399781172007509383004790996974864098331358310421957,...}},...}]},...}, <<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJqaWQiOiJTZWJhc3RpYWFuQGNoYXQuZGV2LnJlenpvbmF0aW9uLmFwcC...">>, <<12,80,80,93,77,192,82,118,51,227,93,211,133,237,4,49,42,143,3,232,59,74,4,147,89,105,118,93,232,...>>, 'RS256') line 64 in p1_server:terminate/7 line 878
07:56:17.554 [error] Supervisor ejabberd_c2s_sup had child undefined started with {ejabberd_c2s,start_link,undefined} at <0.544.0> exit with reason no function clause matching jose_jws_alg_rsa_pkcs1_v1_5:verify({jose_jwk,{jose_jwk_set,[{jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPrivateKey','two-prime',21742263746488030104405521343183166191161927668668832678963882533224849627713697096673808942396166632655001363829479887991333419338499736733930419606265074138668507628098162744273934555101659209242746136159013755169313875220443887183686817460947357037265862949520061234764593811698354537662381041160306313905660977482112950644760943819046932440697293474045016208686816396415168146427032289769084327339610569003142494746410319845820075666726582865656725406634768508632406115803852827347182888089585019913453855785437574949464529944685116786371925939171960320238722474399781172007509383004790996974864098331358310421957,...}},...}]},...}, <<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJqaWQiOiJTZWJhc3RpYWFuQGNoYXQuZGV2LnJlenpvbmF0aW9uLmFwcC...">>, <<12,80,80,93,77,192,82,118,51,227,93,211,133,237,4,49,42,143,3,232,59,74,4,147,89,105,118,93,232,...>>, 'RS256') line 64 in context child_terminated
```

## Bug description

I have ejabberd installed on a ubuntu server. When I use a regular password to login I can connect to the server and start messaging. I switched to JWT for passwords to integrate with an existing backend. When I use the JWT as a password I get the above error as console output when running `ejabberdctl live` and my client (conversejs.org) reports a connection error: 
```
Connection failure

Sorry, we could not connect to the XMPP host with domain: "XXX"
```

Do I need to install an extension / plugin? I didn't install one as I thought it's now integrated in ejabberd. I did enable JWT auth in the config as described in the documentation.

**Comment by licaon-kter** [comment]
Do attach said config too just in case.

**Comment by zinid** [comment]
Show the error from crash.log please.

**Comment by sschimmel** [comment]
This is the output from crash.log:

```
2019-10-23 07:56:17 =ERROR REPORT====
** Generic server <0.544.0> terminating
** Last message in was {'$gen_event',{xmlstreamelement,{xmlel,<<"auth">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-sasl">>},{<<"mechanism">>,<<"PLAIN">>}],[{xmlcdata,<<"c2ViYXN0aWFhbkBjaGF0LmRldi5yZXp6b25hdGlvbi5hcHAAc2ViYXN0aWFhbgBleUpoYkdjaU9pSlNVekkxTmlJc0luUjVjQ0k2SWtwWFZDSjkuZXlKcWFXUWlPaUpUWldKaGMzUnBZV0Z1UUdOb1lYUXVaR1YyTG5KbGVucHZibUYwYVc5dUxtRndjQ0lzSW5OMVlpSTZJbVUzTTJObVl6UmhMVEJqTVdRdE5HTTFOeTFoWldJd0xXRmlOemN6WkRKak1ETmxNU0lzSW1saGRDSTZNVFUzTVRneE5UVXhOeXdpWlhod0lqb3hOakF6TXpVeE5UUTNMQ0pwYzNNaU9pSm9kSFJ3T2k4dmJHOWpZV3hvYjNOME9qZ3dPREFpTENKMWMyVnlibUZ0WlNJNklsTmxZbUZ6ZEdsaFlXNGlMQ0pxZEdraU9pSXpOV0poWmpOa01TMDRZVEJsTFRSa05UQXRPREZtWkMxaU5EZGhObUZtTkRsak1USWlmUS5ERkJRWFUzQVVuWXo0MTNUaGUwRU1TcVBBLWc3U2dTVFdXbDJYZWgwZ0lOU2lOZ002d3p6TGlZZzlZeElFT1V1eXYweVJnOHRUc0ZZYXpneUJobUhYLUxVTWZNRTV4YU9DQXdBVnhXSHl2YjJ3dDdiUUMtX044b0dSZ3JZNGtNdi0wT0dSR2lWc1N4SWNfa0tCLVlXcDY3RDJSc1otYl9zalI0dVBVQ1RySlg0VFhFYTRYMFY3bDBfb1FxUnJfSk5NZG1QckxJeGhsdkhoNlJ0Q21OUnBTN2ZiczhrSHVYdVlhUkoyOGZzb1BIejRfckN2cFFaUGVadmVjU2NWQVpGZGtoR3I0MVVJbk52dFVfdjdXRE53OE0zMjhUOGZ2MW5nVzRnbFVoZjRvX2ItOERJbDlUQlZnN0lscEVWcTZiVU04MUlqdjFLVGp3WnBleURVTmdjanc=">>}]}}}
** When Server state == #{xmlns => <<"jabber:client">>,mgmt_stanzas_req => 0,lserver => <<"chat.dev.rezzonation.app">>,user => <<>>,socket => {socket_state,fast_tls,{tlssock,#Port<0.21>,#Ref<0.3331268381.2066087937.242940>},262144,#Ref<0.3331268381.2066087937.242924>,{state,1000,1000,146,1571817377529902},none},mgmt_resend => if_offline,resource => <<>>,stream_header_sent => true,server => <<"chat.dev.rezzonation.app">>,stream_state => wait_for_sasl_request,mgmt_stanzas_out => 0,stream_timeout => {30000,-576460738476},tls_required => true,pres_a => {0,nil},mgmt_max_timeout => 300000,mgmt_ack_timeout => 60000,mgmt_state => inactive,mgmt_timeout => 300000,socket_monitor => #Ref<0.3331268381.2065956865.242925>,mgmt_queue_type => ram,tls_options => [compression_none],access => c2s,stream_id => <<"14591792612872742335">>,stream_direction => in,owner => <0.544.0>,csi_queue => {0,#{}},codec_options => [ignore_els],mgmt_stanzas_in => 0,stream_authenticated => false,mgmt_max_queue => 5000,mod => ejabberd_c2s,zlib => false,ip => {{0,0,0,0,0,65535,35529,23390},36194},stream_restarted => true,lang => <<"en">>,stream_compressed => false,tls_verify => false,stream_encrypted => true,tls_enabled => false,csi_state => active,stream_version => {1,0},shaper => c2s_shaper}
** Reason for termination ==
** {function_clause,[{jose_jws_alg_rsa_pkcs1_v1_5,verify,[{jose_jwk,{jose_jwk_set,[{jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPrivateKey','two-prime',21742263746488030104405521343183166191161927668668832678963882533224849627713697096673808942396166632655001363829479887991333419338499736733930419606265074138668507628098162744273934555101659209242746136159013755169313875220443887183686817460947357037265862949520061234764593811698354537662381041160306313905660977482112950644760943819046932440697293474045016208686816396415168146427032289769084327339610569003142494746410319845820075666726582865656725406634768508632406115803852827347182888089585019913453855785437574949464529944685116786371925939171960320238722474399781172007509383004790996974864098331358310421957,65537,16532042769961876743865845851557189643071565368953496041291609234412022317900254111745262175831755735515266612178626451903262690579002988860112923536619027030198652244127556126656067667879608186444062521279215915400418832567635988619214817396492495779193928353456577070216659884104435712968121397731058550658894932453904371171158060498084004941470684151424906146706939814669143990961239678003726011856947192861615182531349861154596255759310691144020220983193205344264153870164938357951180141287614822733540940313633954600924785154667128441506038285552661344274432401436929010751792359541394901361915678539301637364449,174219236919982577766531195208427115181985376844835104447946348032331530289473081748458342833784532776172655667440033538712256531058480239564166898352921362733466383405393967632653378953746179913626308052679701836305430875142648830618259350007567266962986577703494712508936425720931021814286564755409965353543,124798295130141500848445958584623646089151666955878765646545545373290747305749030276802906181280599989779394835169598987939918716300130931084266194597204774590546487314987091271631513479103958046703599158856996514170343958741785730055696788227951278518281323470808522076028891878811057787085242990978618628499,95197599727505013845571946100813028087676859334275141147525926871627087465039909219128762610749010532180584587434503273796857021739402716923758804902825379874699248609648952117017107186972312569798308095801040671070292889050969639642804607826433791556090640922530008234394652509151484233202865106062014881299,35285600634168820829786282749791356974411101952979744685147152841403749753200932771246133505334841659072160555040552195653244637579404399850335727694069067445303056440586398542248377935636302281236823358005708918741725644376234639637640744707019503348394844498742419001004247471113552661774105507161356229093,92504444044244523650180240792298361816613169539192923152359388350651122022309720695994231293306303381472555048182213773074286457258067871834402598608560341487019095848056517451537654565450795253312178421475437745395630994536919544413711086382141978123550803600059439028827896983192232241884214086208711186167,asn1_NOVALUE}},#{<<"alg">> => <<"RS256">>}}]},undefined,#{}},<<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJqaWQiOiJTZWJhc3RpYWFuQGNoYXQuZGV2LnJlenpvbmF0aW9uLmFwcCIsInN1YiI6ImU3M2NmYzRhLTBjMWQtNGM1Ny1hZWIwLWFiNzczZDJjMDNlMSIsImlhdCI6MTU3MTgxNTUxNywiZXhwIjoxNjAzMzUxNTQ3LCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAiLCJ1c2VybmFtZSI6IlNlYmFzdGlhYW4iLCJqdGkiOiIzNWJhZjNkMS04YTBlLTRkNTAtODFmZC1iNDdhNmFmNDljMTIifQ">>,<<12,80,80,93,77,192,82,118,51,227,93,211,133,237,4,49,42,143,3,232,59,74,4,147,89,105,118,93,232,116,128,131,82,136,216,12,235,12,243,46,38,32,245,140,72,16,229,46,202,253,50,70,15,45,78,193,88,107,56,50,6,25,135,95,226,212,49,243,4,231,22,142,8,12,0,87,21,135,202,246,246,194,222,219,64,47,191,55,202,6,70,10,216,226,67,47,251,67,134,68,104,149,177,44,72,115,249,10,7,230,22,167,174,195,217,27,25,249,191,236,141,30,46,61,64,147,172,149,248,77,113,26,225,125,21,238,93,63,161,10,145,175,242,77,49,217,143,172,178,49,134,91,199,135,164,109,10,99,81,165,46,223,110,207,36,30,229,238,97,164,73,219,199,236,160,241,243,227,250,194,190,148,25,61,230,111,121,196,156,84,6,69,118,72,70,175,141,84,34,115,111,181,79,239,237,96,205,195,195,55,219,196,252,126,253,103,129,110,32,149,72,95,226,143,219,251,192,200,151,212,193,86,14,200,150,145,21,171,166,212,51,205,72,142,253,74,78,60,25,165,236,131,80,216,28,143>>,'RS256'],[{file,"src/jose_jws_alg_rsa_pkcs1_v1_5.erl"},{line,64}]},{jose_jws,verify,2,[{file,"src/jose_jws.erl"},{line,379}]},{jose_jwt,verify,2,[{file,"src/jose_jwt.erl"},{line,185}]},{ejabberd_auth_jwt,check_jwt_token,3,[{file,"src/ejabberd_auth_jwt.erl"},{line,89}]},{ejabberd_auth_jwt,check_password,4,[{file,"src/ejabberd_auth_jwt.erl"},{line,67}]},{ejabberd_auth,db_check_password,7,[{file,"src/ejabberd_auth.erl"},{line,676}]},{ejabberd_auth,'-check_password_with_authmodule/6-fun-0-',8,[{file,"src/ejabberd_auth.erl"},{line,243}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]}]}
2019-10-23 07:56:17 =CRASH REPORT====
  crasher:
    initial call: xmpp_stream_in:init/1
    pid: <0.544.0>
    registered_name: []
    exception exit: {{function_clause,[{jose_jws_alg_rsa_pkcs1_v1_5,verify,[{jose_jwk,{jose_jwk_set,[{jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPrivateKey','two-prime',21742263746488030104405521343183166191161927668668832678963882533224849627713697096673808942396166632655001363829479887991333419338499736733930419606265074138668507628098162744273934555101659209242746136159013755169313875220443887183686817460947357037265862949520061234764593811698354537662381041160306313905660977482112950644760943819046932440697293474045016208686816396415168146427032289769084327339610569003142494746410319845820075666726582865656725406634768508632406115803852827347182888089585019913453855785437574949464529944685116786371925939171960320238722474399781172007509383004790996974864098331358310421957,65537,16532042769961876743865845851557189643071565368953496041291609234412022317900254111745262175831755735515266612178626451903262690579002988860112923536619027030198652244127556126656067667879608186444062521279215915400418832567635988619214817396492495779193928353456577070216659884104435712968121397731058550658894932453904371171158060498084004941470684151424906146706939814669143990961239678003726011856947192861615182531349861154596255759310691144020220983193205344264153870164938357951180141287614822733540940313633954600924785154667128441506038285552661344274432401436929010751792359541394901361915678539301637364449,174219236919982577766531195208427115181985376844835104447946348032331530289473081748458342833784532776172655667440033538712256531058480239564166898352921362733466383405393967632653378953746179913626308052679701836305430875142648830618259350007567266962986577703494712508936425720931021814286564755409965353543,124798295130141500848445958584623646089151666955878765646545545373290747305749030276802906181280599989779394835169598987939918716300130931084266194597204774590546487314987091271631513479103958046703599158856996514170343958741785730055696788227951278518281323470808522076028891878811057787085242990978618628499,95197599727505013845571946100813028087676859334275141147525926871627087465039909219128762610749010532180584587434503273796857021739402716923758804902825379874699248609648952117017107186972312569798308095801040671070292889050969639642804607826433791556090640922530008234394652509151484233202865106062014881299,35285600634168820829786282749791356974411101952979744685147152841403749753200932771246133505334841659072160555040552195653244637579404399850335727694069067445303056440586398542248377935636302281236823358005708918741725644376234639637640744707019503348394844498742419001004247471113552661774105507161356229093,92504444044244523650180240792298361816613169539192923152359388350651122022309720695994231293306303381472555048182213773074286457258067871834402598608560341487019095848056517451537654565450795253312178421475437745395630994536919544413711086382141978123550803600059439028827896983192232241884214086208711186167,asn1_NOVALUE}},#{<<"alg">> => <<"RS256">>}}]},undefined,#{}},<<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJqaWQiOiJTZWJhc3RpYWFuQGNoYXQuZGV2LnJlenpvbmF0aW9uLmFwcCIsInN1YiI6ImU3M2NmYzRhLTBjMWQtNGM1Ny1hZWIwLWFiNzczZDJjMDNlMSIsImlhdCI6MTU3MTgxNTUxNywiZXhwIjoxNjAzMzUxNTQ3LCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAiLCJ1c2VybmFtZSI6IlNlYmFzdGlhYW4iLCJqdGkiOiIzNWJhZjNkMS04YTBlLTRkNTAtODFmZC1iNDdhNmFmNDljMTIifQ">>,<<12,80,80,93,77,192,82,118,51,227,93,211,133,237,4,49,42,143,3,232,59,74,4,147,89,105,118,93,232,116,128,131,82,136,216,12,235,12,243,46,38,32,245,140,72,16,229,46,202,253,50,70,15,45,78,193,88,107,56,50,6,25,135,95,226,212,49,243,4,231,22,142,8,12,0,87,21,135,202,246,246,194,222,219,64,47,191,55,202,6,70,10,216,226,67,47,251,67,134,68,104,149,177,44,72,115,249,10,7,230,22,167,174,195,217,27,25,249,191,236,141,30,46,61,64,147,172,149,248,77,113,26,225,125,21,238,93,63,161,10,145,175,242,77,49,217,143,172,178,49,134,91,199,135,164,109,10,99,81,165,46,223,110,207,36,30,229,238,97,164,73,219,199,236,160,241,243,227,250,194,190,148,25,61,230,111,121,196,156,84,6,69,118,72,70,175,141,84,34,115,111,181,79,239,237,96,205,195,195,55,219,196,252,126,253,103,129,110,32,149,72,95,226,143,219,251,192,200,151,212,193,86,14,200,150,145,21,171,166,212,51,205,72,142,253,74,78,60,25,165,236,131,80,216,28,143>>,'RS256'],[{file,"src/jose_jws_alg_rsa_pkcs1_v1_5.erl"},{line,64}]},{jose_jws,verify,2,[{file,"src/jose_jws.erl"},{line,379}]},{jose_jwt,verify,2,[{file,"src/jose_jwt.erl"},{line,185}]},{ejabberd_auth_jwt,check_jwt_token,3,[{file,"src/ejabberd_auth_jwt.erl"},{line,89}]},{ejabberd_auth_jwt,check_password,4,[{file,"src/ejabberd_auth_jwt.erl"},{line,67}]},{ejabberd_auth,db_check_password,7,[{file,"src/ejabberd_auth.erl"},{line,676}]},{ejabberd_auth,'-check_password_with_authmodule/6-fun-0-',8,[{file,"src/ejabberd_auth.erl"},{line,243}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]}]},[{p1_server,terminate,7,[{file,"src/p1_server.erl"},{line,878}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,249}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.112.0>]
    message_queue_len: 0
    messages: []
    links: [<0.539.0>]
    dictionary: [{already_terminated,true},{'$internal_queue_len',0},{rand_seed,{#{bits => 58,jump => #Fun<rand.8.10897371>,next => #Fun<rand.5.10897371>,type => exrop,uniform => #Fun<rand.6.10897371>,uniform_n => #Fun<rand.7.10897371>,weak_low_bits => 1},[79442829740612374|173153225720144442]}}]
    trap_exit: false
    status: running
    heap_size: 1598
    stack_size: 27
    reductions: 10174
  neighbours:
2019-10-23 07:56:17 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {function_clause,[{jose_jws_alg_rsa_pkcs1_v1_5,verify,[{jose_jwk,{jose_jwk_set,[{jose_jwk,undefined,{jose_jwk_kty_rsa,{'RSAPrivateKey','two-prime',21742263746488030104405521343183166191161927668668832678963882533224849627713697096673808942396166632655001363829479887991333419338499736733930419606265074138668507628098162744273934555101659209242746136159013755169313875220443887183686817460947357037265862949520061234764593811698354537662381041160306313905660977482112950644760943819046932440697293474045016208686816396415168146427032289769084327339610569003142494746410319845820075666726582865656725406634768508632406115803852827347182888089585019913453855785437574949464529944685116786371925939171960320238722474399781172007509383004790996974864098331358310421957,65537,16532042769961876743865845851557189643071565368953496041291609234412022317900254111745262175831755735515266612178626451903262690579002988860112923536619027030198652244127556126656067667879608186444062521279215915400418832567635988619214817396492495779193928353456577070216659884104435712968121397731058550658894932453904371171158060498084004941470684151424906146706939814669143990961239678003726011856947192861615182531349861154596255759310691144020220983193205344264153870164938357951180141287614822733540940313633954600924785154667128441506038285552661344274432401436929010751792359541394901361915678539301637364449,174219236919982577766531195208427115181985376844835104447946348032331530289473081748458342833784532776172655667440033538712256531058480239564166898352921362733466383405393967632653378953746179913626308052679701836305430875142648830618259350007567266962986577703494712508936425720931021814286564755409965353543,124798295130141500848445958584623646089151666955878765646545545373290747305749030276802906181280599989779394835169598987939918716300130931084266194597204774590546487314987091271631513479103958046703599158856996514170343958741785730055696788227951278518281323470808522076028891878811057787085242990978618628499,95197599727505013845571946100813028087676859334275141147525926871627087465039909219128762610749010532180584587434503273796857021739402716923758804902825379874699248609648952117017107186972312569798308095801040671070292889050969639642804607826433791556090640922530008234394652509151484233202865106062014881299,35285600634168820829786282749791356974411101952979744685147152841403749753200932771246133505334841659072160555040552195653244637579404399850335727694069067445303056440586398542248377935636302281236823358005708918741725644376234639637640744707019503348394844498742419001004247471113552661774105507161356229093,92504444044244523650180240792298361816613169539192923152359388350651122022309720695994231293306303381472555048182213773074286457258067871834402598608560341487019095848056517451537654565450795253312178421475437745395630994536919544413711086382141978123550803600059439028827896983192232241884214086208711186167,asn1_NOVALUE}},#{<<"alg">> => <<"RS256">>}}]},undefined,#{}},<<"eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJqaWQiOiJTZWJhc3RpYWFuQGNoYXQuZGV2LnJlenpvbmF0aW9uLmFwcCIsInN1YiI6ImU3M2NmYzRhLTBjMWQtNGM1Ny1hZWIwLWFiNzczZDJjMDNlMSIsImlhdCI6MTU3MTgxNTUxNywiZXhwIjoxNjAzMzUxNTQ3LCJpc3MiOiJodHRwOi8vbG9jYWxob3N0OjgwODAiLCJ1c2VybmFtZSI6IlNlYmFzdGlhYW4iLCJqdGkiOiIzNWJhZjNkMS04YTBlLTRkNTAtODFmZC1iNDdhNmFmNDljMTIifQ">>,<<12,80,80,93,77,192,82,118,51,227,93,211,133,237,4,49,42,143,3,232,59,74,4,147,89,105,118,93,232,116,128,131,82,136,216,12,235,12,243,46,38,32,245,140,72,16,229,46,202,253,50,70,15,45,78,193,88,107,56,50,6,25,135,95,226,212,49,243,4,231,22,142,8,12,0,87,21,135,202,246,246,194,222,219,64,47,191,55,202,6,70,10,216,226,67,47,251,67,134,68,104,149,177,44,72,115,249,10,7,230,22,167,174,195,217,27,25,249,191,236,141,30,46,61,64,147,172,149,248,77,113,26,225,125,21,238,93,63,161,10,145,175,242,77,49,217,143,172,178,49,134,91,199,135,164,109,10,99,81,165,46,223,110,207,36,30,229,238,97,164,73,219,199,236,160,241,243,227,250,194,190,148,25,61,230,111,121,196,156,84,6,69,118,72,70,175,141,84,34,115,111,181,79,239,237,96,205,195,195,55,219,196,252,126,253,103,129,110,32,149,72,95,226,143,219,251,192,200,151,212,193,86,14,200,150,145,21,171,166,212,51,205,72,142,253,74,78,60,25,165,236,131,80,216,28,143>>,'RS256'],[{file,"src/jose_jws_alg_rsa_pkcs1_v1_5.erl"},{line,64}]},{jose_jws,verify,2,[{file,"src/jose_jws.erl"},{line,379}]},{jose_jwt,verify,2,[{file,"src/jose_jwt.erl"},{line,185}]},{ejabberd_auth_jwt,check_jwt_token,3,[{file,"src/ejabberd_auth_jwt.erl"},{line,89}]},{ejabberd_auth_jwt,check_password,4,[{file,"src/ejabberd_auth_jwt.erl"},{line,67}]},{ejabberd_auth,db_check_password,7,[{file,"src/ejabberd_auth.erl"},{line,676}]},{ejabberd_auth,'-check_password_with_authmodule/6-fun-0-',8,[{file,"src/ejabberd_auth.erl"},{line,243}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]}]}
     Offender:   [{pid,<0.544.0>},{id,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,5000},{child_type,worker}]
```

**Comment by sschimmel** [comment]
And the host config from ejabberd.yml

```
host_config:
  XXX:
    auth_method: jwt
    jwt_key: /opt/ejabberd/conf/jwt-dev.key
```

**Comment by zinid** [comment]
BTW, I hope your private key is not from production, because you disclosed it here ;)

**Comment by sschimmel** [comment]
Hmm, I was afraid I'd do that. But no worries, it's a development key. :-)

**Comment by zinid** [comment]
I know very little about JWT, but according to the crash report, maybe you put some wrong JWT structure into your key file? Because I see that `kty` attribute is missing, and thus the library crashes.

**Comment by sschimmel** [comment]
The "kty" attribute is part of the JWK spec and the attribute is present the JWK used on the ejabberd server.

JWK used:
```
{
  "keys": [
    {
      "kty": "RSA",
      "d": "gvWBVItA0cLrSkIqQwfwoTAXe33t5Z47-7p8VzvxLN_jPTjKs68Ryk-TqNOG1SZxW2fFdTV1Z3RC6diME...b2nq0ol9KcxIJl9oMTggv3YRS-qZbkh2wvE...ia4Q",
      "e": "AQAB",
      "alg": "RS256",
      "n": "rDteIAn5KZHCg...I7-JadMwvAFsvJ_VxQ"
    }
  ]
}
```

I can't find any reference to the attribute in the JWT specs.

**Comment by zinid** [comment]
> When I change the JWK to something that's not valid ejabberd will complain on start but with this key it doens't throw any errors or warnings so I guess it was able to successfully parse the JWK and accepted it as valid.

Yes, I know, that's the mysterious part: looks like the library is doing something wrong. ejabberd checks the key on configuration load using this library and then this exact library crashes on this exact key, LOL.

**Comment by zinid** [comment]
But I will wait for @alexeyshch input.

**Comment by zinid** [comment]
@sschimmel meanwhile, if you just put
```json
{
      "kty": "RSA",
      "d": "gvWBVItA0cLrSkIqQwfwoTAXe33t5Z47-7p8VzvxLN_jPTjKs68Ryk-TqNOG1SZxW2fFdTV1Z3RC6diME...b2nq0ol9KcxIJl9oMTggv3YRS-qZbkh2wvE...ia4Q",
      "e": "AQAB",
      "alg": "RS256",
      "n": "rDteIAn5KZHCg...I7-JadMwvAFsvJ_VxQ"
}
```
in the key file, does it work?

**Comment by alexeyshch** [comment]
I can reproduce the error if the key file contains JWK set (i.e. it has "keys" attribute with a list of JWKs), I guess it's not supported by the library.

**Comment by zinid** [comment]
@alexeyshch anyway, ejabberd should not crash in runtime and should reject the key file at configuration load.

**Comment by alexeyshch** [comment]
Should be fixed in c7470f510762b3ff49dde74a6c780a0cf0ebed8a.
